swarm/src/behaviour/external_addresses.rs

```diff
@@ -29,18 +18,40 @@ impl ExternalAddresses {
         self.addresses.iter()
     }
 
+    pub fn as_slice(&self) -> &[Multiaddr] {
+        self.addresses.as_slice()
+    }
+
     /// Feed a [`FromSwarm`] event to this struct.
     ///
     /// Returns whether the event changed our set of external addresses.
     pub fn on_swarm_event<THandler>(&mut self, event: &FromSwarm<THandler>) -> bool {
         match event {
             FromSwarm::ExternalAddrConfirmed(ExternalAddrConfirmed { addr }) => {
-                if self.addresses.len() < self.limit {
-                    return self.addresses.insert((*addr).clone());
+                if self.addresses.contains(addr) {
+                    return false;
+                }
+
+                if self.addresses.len() >= MAX_LOCAL_EXTERNAL_ADDRS {
+                    // If we hit the limit, we can add the address in O(1):
+                    self.addresses.push((*addr).clone());
+                    self.addresses.swap_remove(0);
```

Comment:
Can you add a `log::debug` line here? I would expect this to be hard to debug otherwise.

Reply:
Ah, this isn't as clean as I thought it is because we are not removing the oldest entry ...